common/lib/host_list_provider/monitoring/cluster_topology_monitor.ts

```diff
@@ -0,0 +1,482 @@
+    refreshRateMs: number,
+    highRefreshRateMs: number
+  ) {
+    this.clusterId = clusterId;
+    this.topologyMap = topologyMap;
+    this.initialHostInfo = initialHostSpec;
+    this._pluginService = pluginService;
+    this._hostListProvider = hostListProvider;
+    this._properties = props;
+    this.refreshRateMs = refreshRateMs;
+    this.highRefreshRateMs = highRefreshRateMs;
+    this.runPromise = this.run();
+  }
+
+  get hostListProvider(): MonitoringRdsHostListProvider {
+    return this._hostListProvider;
+  }
+  get pluginService(): PluginService {
+    return this._pluginService;
+  }
+  get properties(): Map<string, any> {
+    return this._properties;
+  }
+
+  async close(): Promise<void> {
+    this.stopMonitoring = true;
+    this.hostMonitorsStop = true;
+    await Promise.all(this.runPromise ? this.untrackedPromises.concat(this.runPromise) : this.untrackedPromises);
+    this.untrackedPromises = [];
+    this.hostMonitors.clear();
+  }
+
+  async forceMonitoringRefresh(shouldVerifyWriter: boolean, timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts = this.topologyMap.get(this.clusterId);
+    if (currentHosts) {
+      return currentHosts;
+    }
+    if (shouldVerifyWriter) {
+      const client = this.monitoringClient;
+      this.monitoringClient = null;
+      this.isVerifiedWriterConnection = false;
+      await this.closeConnection(client);
+    }
+
+    return await this.waitTillTopologyGetsUpdated(timeoutMs);
+  }
+
+  async forceRefresh(client: any, timeoutMs: number): Promise<HostInfo[]> {
+    if (this.isVerifiedWriterConnection) {
+      return await this.waitTillTopologyGetsUpdated(timeoutMs);
+    }
+
+    // Otherwise use provided unverified connection to update topology
+    return await this.fetchTopologyAndUpdateCache(client);
+  }
+
+  async waitTillTopologyGetsUpdated(timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts: HostInfo[] = this.topologyMap.get(this.clusterId);
+    let latestHosts: HostInfo[];
+
+    // Notify monitor (that might be in delay) that topology should be updated.
+    this.requestToUpdateTopology = true;
+
+    if (timeoutMs === 0) {
+      return currentHosts;
+    }
+
+    let timeoutId: any;
+    const timeoutTask: Promise<void> = new Promise((resolve, reject) => {
+      timeoutId = setTimeout(() => {
+        reject("Topology update timed out.");
+      }, timeoutMs);
+    });
+
+    return await Promise.race([timeoutTask, await this.topologyUpdated])
+      .then((result) => {
+        latestHosts = this.topologyMap.get(this.clusterId);
+        return latestHosts === currentHosts ? null : latestHosts;
+      })
+      .catch((error: any) => {
+        logger.debug(`Could not update topology: ${error}.`);
+        if (JSON.stringify(error).includes("Topology update timed out.")) {
+          throw new AwsWrapperError(`ClusterTopologyMonitor topology update timed out in ${timeoutMs} ms.`);
+        }
+        return null;
+      })
```

Comment:
I think the goal of this method to keep calling `this.topologyMap.get(this.clusterId)` until either current hosts != latest hosts or time out has exceeded. This code seems to return immediately after one call.

Also should not be calling await here: `Promise.race([timeoutTask, await this.topologyUpdated])`.